src/core/net/ip6.cpp

```diff
@@ -977,6 +982,17 @@ Error Ip6::PassToHost(Message           &aMessage,
     // Do not pass IPv6 packets that exceed kMinimalMtu.
     VerifyOrExit(aMessage.GetLength() <= kMinimalMtu, error = kErrorDrop);
 
+#if OPENTHREAD_CONFIG_IP6_CORE_FIREWALL_ENABLE
+    // When the packet is classified as 'receive' or it has a multicast destination we should always forward it to host.
+    // When the packet's source is mesh-local and the destination is not on the host, we should not forward it to host.
+    VerifyOrExit(aReceive || aMessageInfo.GetSockAddr().IsMulticast() ||
+                     !Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetPeerAddr()) ||
+                     otPlatHostHasAddress(&aMessageInfo.GetSockAddr()),
```

Comment:
should we force the destination is mesh-local?

Reply:
I'm not sure if I get your point. Could you elaborate?

I think any destination on the host should be allowed here.

Reply:
Do we have use cases to send to non-mesh-local host addresses with mesh-local source address? If we only allow mesh-local destination address or addresses on the Thread's network interface, we probably can eliminate `otPlatHostHasAddress()`.

Reply:
I'm not aware of such a use case but this kind of packets should be allowed.

Reply:
If a packet is sent using a non-mesh-local source address from host, there may be no route back to the source, so I don't think we should support it, especially when it adds the platform dependency `otPlatHostHasAddress()`. Thoughts?

Reply:
Updated the code to avoid checking all addresses on host.